### PR TITLE
Implement failpoint counter

### PR DESCRIPTION
**Pull request by ZhouJianMS**
This is PR based out of @ramil600 PR https://github.com/etcd-io/gofail/pull/37 and @pchan PR https://github.com/etcd-io/gofail/pull/47 and is intended to address https://github.com/etcd-io/etcd/issues/14729.

New changes include:
1. Update the design doc and readme [(#47(comment1))](https://github.com/etcd-io/gofail/pull/47#issuecomment-1570284395)
2. Move tests to terms_test.go [(#47(comment2))](https://github.com/etcd-io/gofail/pull/47#issuecomment-1584452870)

cc: @ramil600 @pchan @serathius @ahrtr @lavacat

### CODE REVIEW

File: runtime/http.go [hunk 1]
```diff
@@ -87,12 +88,24 @@ func (*httpHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			sort.Strings(fps)
 			lines := make([]string, len(fps))
 			for i := range lines {
-				s, _ := status(fps[i])
+				s, _, _ := status(fps[i])
 				lines[i] = fps[i] + "=" + s
 			}
 			w.Write([]byte(strings.Join(lines, "\n") + "\n"))
+		} else if strings.HasSuffix(key, "/execution-count") {
```
Comment (serathius) [inline runtime/http.go:95]: Don't think the change from "/count" to "/execution-count" was needed.

**Review by serathius — APPROVED** [review]
Let's merge it and start using in robustness tests so we can iterate.

File: runtime/http.go (hunk 1 above)
Comment (ZhouJianMS) [reply runtime/http.go]: Recovered to "/count"

**Comment by ZhouJianMS** [comment]
> Please fix the failures

lint failures fixed, please review

File: doc/design.md [hunk 2]
```diff
@@ -69,6 +69,16 @@ Similarly, you can set multiple failpoints using endpoint `/failpoints`,
 curl http://127.0.0.1:22381/failpoints -X PUT -d'failpoint1=return("hello");failpoint2=sleep(10)'
 ```
 
+You can get the execution count of the failpoint in the dynamic way,
+```
+$curl http://127.0.0.1:1234/SomeFuncString/execution-count -XGET
```
Comment (serathius) [inline doc/design.md:74]:
```suggestion
$curl http://127.0.0.1:1234/SomeFuncString/count -XGET
```

File: README.md [hunk 3]
```diff
@@ -86,6 +86,12 @@ List the failpoints,
 $ curl http://127.0.0.1:1234/SomeFuncString=return("hello")
 ```
 
+Retrieve the execution count of a failpoint,
+
+```sh
+$curl http://127.0.0.1:1234/SomeFuncString/execution-count -XGET
```
Comment (serathius) [inline README.md:92]:
```suggestion
$curl http://127.0.0.1:1234/SomeFuncString/count -XGET
```

File: runtime/terms_test.go [hunk 4]
```diff
@@ -71,3 +71,45 @@ func TestTermsTypes(t *testing.T) {
...
+		ter, err := newTerms("test", tt.failpointTerm)
+		if err != nil {
+			t.Fatal(err)
+		}
+		for i := 0; i < tt.runAfterEnabling; i++ {
+			_ = ter.eval()
+		}
+
+		if ter.counter != tt.wantCount {
+			t.Fatalf("counter is not properly incremented")
```
Comment (serathius) [inline runtime/terms_test.go:112]:
```suggestion
			t.Eatalf("counter is not properly incremented, got: %d, want: %d", ter.counter, tt.wantCount)
```

File: README.md (hunk 3 above)
Comment (ZhouJianMS) [reply README.md]: Done. Thanks for correction!

**Comment by serathius** [comment]
cc @ahrtr

File: runtime/terms_test.go [hunk 5]
```diff
@@ -71,3 +71,45 @@ func TestTermsTypes(t *testing.T) {
...
+		wantCount        int
+	}{
+		{
+			failpointTerm:    `10*sleep(10)->1*return("abc")`,
+			runAfterEnabling: 12,
+			// This example tests mods which allows users to restrict the
+			// number of failpoint actions as against their callsite executions.
+			// This is the reason why wantCount < runAfterEnabling
+			// In a real world example you can hit a code spot a million times but
+			// using mods restrict the associated fallpoint actions to run twice.
```
Comment (ahrtr) [inline runtime/terms_test.go:88]: It's a good test case, but the comment isn't clear to me.
```suggestion
			// Note the chain of terms is allowed to be executed 11 time at most, 
			// including 10 times for the first term `10*sleep(10)` and 1 time for
			// the second term `1*return("abc")`. So it's only evaluated 11 times
			// even it's triggered 12 times.
```

**Review by ahrtr — COMMENTED** [review]
Overall looks good with two minor comments. Thanks.

**Review by jmhbnz — APPROVED** [review]
Great work @ZhouJianMS.  Looks good, one minor suggestion below.

File: runtime/http.go [hunk 6]
```diff
@@ -87,12 +88,24 @@ func (*httpHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
-				s, _ := status(fps[i])
+				s, _, _ := status(fps[i])
 				lines[i] = fps[i] + "=" + s
 			}
 			w.Write([]byte(strings.Join(lines, "\n") + "\n"))
+		} else if strings.HasSuffix(key, "/count") {
+			fp := key[:len(key)-len("/count")]
+			_, count, err := status(fp)
+			if err != nil {
+				if err == ErrNoExist {
```
Comment (jmhbnz) [inline runtime/http.go:99]: Nit - Suggest [`errors.Is`](https://github.com/golang/go/wiki/ErrorValueFAQ#how-should-i-change-my-error-handling-code-to-work-with-the-new-features).

```suggestion
				if errors.Is(err, io.ErrNoExist) {
```

File: runtime/terms_test.go (hunk 5 above)
Comment (ZhouJianMS) [reply runtime/terms_test.go]: Done. Thanks for your suggestions!

File: runtime/http.go (hunk 6 above)
Comment (ZhouJianMS) [reply runtime/http.go]: Done. Thanks for your suggestion!

**Review by ahrtr — APPROVED** [review]
lgtm

Thanks.

**Pull request merged**